### PR TITLE
History manager

### PR DESCRIPTION
**Pull request by JunkuiZhang**
While working on implementing `add_recent_documents` for Windows, I found that the process is significantly more complex compared to macOS. On macOS, simply registering the `add_recent_documents` function is enough, as the system handles everything automatically.  

On Windows, however, there are two cases to consider:  
- **Files opened by the app**: These appear in the "Recent" section (as shown in the screenshot, "test.txt") and are managed automatically by Windows (by setting windows registry), similar to macOS.  

屏幕截图 2025-03-10 230738

- **Folders opened by the app**: This is more complicated because Windows does not handle it automatically, requiring the application to track opened folders manually.  

To address this, this PR introduces a `History Manager` along with `HistoryManagerEvent::Update` and `HistoryManagerEvent::Delete` events to simplify the process of managing recently opened folders.

https://github.com/user-attachments/assets/a2581c15-7653-4faf-96b0-7c48ab1dcc8d

Release Notes:

- N/A

### CODE REVIEW

**Comment by JunkuiZhang** [comment]
This PR moves away from the WinRT-based `JumpList` API used in #26010 due to its limitations in flexibility. In particular, when setting icons for `Recent Folders`, `JumpList` can only use image resources from the `appx` package, making it impossible to retrieve folder icons dynamically from `explorer.exe` as VS Code does. To address this, this PR switches to using the Win32 API instead.

**Review by SomeoneToIgnore — COMMENTED** [review]
So, this version will be able to unite Windows "icon context menu" with the DB state, and recent projects modal, if I get it right?

That's great, as it's what people wanted for macOS too — yet there, we only add opened projects on top.
Not sure if we should do it here (would be nice though!), but definitely should somehow keep the API of gpui conslidated, if possible.

Besides gpui, 2 main notes I'd love to highlight:
* DB access in detached async tasks, potentially parallel, might be a bit overly racy
* the usage of history manager: it's `Entity` is global, but is only modified through emitting events; the manager itself and existing `RecentProjects` have very similar goals, so maybe their code should be together?

File: crates/workspace/src/workspace.rs [hunk 1]
```diff
@@ -4502,22 +4516,38 @@ impl Workspace {
         cx.notify();
     }
 
-    fn serialize_workspace(&mut self, window: &mut Window, cx: &mut Context<Self>) {
+    /// Set the parameter `update` to `true` if the action changes the history of the workspace.
+    /// Typically, there are three cases that change the history:
+    /// 1. A brand new workspace is opened, or an existing workspace has been deleted.
+    /// 2. An existing workspace is loaded, this will change the order of the workspaces in the history.
+    /// 3. `AddFolderToProject` or `RemoveFromProject` action is triggered.
+    fn serialize_workspace(&mut self, window: &mut Window, cx: &mut Context<Self>, update: bool) {
```
[the remark this answers is not in the document]
Comment (JunkuiZhang) [reply crates/workspace/src/workspace.rs:4524]: Got it!

File: crates/workspace/src/workspace.rs [hunk 2]
```diff
@@ -4657,6 +4687,16 @@ impl Workspace {
                 session_id: self.session_id.clone(),
                 window_id: Some(window.window_handle().window_id().as_u64()),
             };
+            if update {
+                if let Some(manager) = HistoryManager::global(cx) {
+                    return window.spawn(cx, |mut cx| async move {
+                        persistence::DB.save_workspace(serialized_workspace).await;
+                        manager
+                            .update(&mut cx, |_, cx| cx.emit(HistoryManagerEvent::Update))
```
[the remark this answers is not in the document]
Comment (JunkuiZhang) [reply crates/workspace/src/workspace.rs:4695]: Currently, I’ve only found this one place where `DB.save_workspace()` is called, so the `update` only needs to be handled here. Similarly, it looks like `DB.delete_workspace()` is only called in `recent_projects`, so there’s just one place where `delete` is needed.  

For better scalability, I chose to use an `Event`-based approach. Although we only have two cases—`update` and `delete`—for now, if more components affect `History` in the future, they can simply emit the corresponding `Event`, and we can centralize all the event-handling logic in one place.

File: crates/workspace/src/history_manager.rs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+use crate::WORKSPACE_DB;
+
+pub fn init(cx: &mut App) {
+    let manager = cx.new(|_| HistoryManager::new());
+    HistoryManager::set_global(manager.clone(), cx);
+    cx.subscribe(&manager, |manager, event, cx| match event {
+        HistoryManagerEvent::Update => perform_update(cx, manager),
+        HistoryManagerEvent::Delete(DeleteSource::User) => {
+            manager.update(cx, |_, cx| cx.emit(HistoryManagerEvent::Update));
```
[the remark this answers is not in the document]
Comment (JunkuiZhang) [reply crates/workspace/src/history_manager.rs:13]: > We have a `&mut history manager` hidden under `_` here, which we can do `perform_update` on, instead of raising the event?
> 
> Both seem to do `perform_update(cx, manager)` so not sure why a separation and event emitting is needed.

On Windows, those two events are the same. I don't know if it's the case on Linux, so I just leave it separate events.

> We might also add the last `DeleteSource` variant to cover all match branches and help future changes to enum with compiler errors here, on the handler side.

Got it!

File: crates/workspace/src/history_manager.rs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+                .into_iter()
+                .filter_map(|(query, id)| {
+                    if user_removed.contains(&query) {
+                        Some(id)
+                    } else {
+                        None
+                    }
+                })
+                .collect::<Vec<_>>();
+            for id in deleted_ids.iter() {
```
[the remark this answers is not in the document]
Comment (JunkuiZhang) [reply crates/workspace/src/history_manager.rs:58]: > So now, we have workspace serialization and this task, both are called on "random" events from the outside and run to the end due to `.detach()`.
> 
> Seems a bit fragile even for the workspace serializations alone tbh, but here we remove from database in a loop, which makes it a bit more interesting.
> 
> I would propose to check with someone else who knows workspace DB caveats better.

Good point! What I can confirm is that this only gets triggered when the user manually deletes an entry via right-click, and even then, only when we next call `list.BeginList()`. This scenario is quite rare.  

> From my side, I would at least make the `delete from WORKSPACE where ID in (?, ?, ..., ?)` call instead, if many workspaces are expected to be removed.
> 
> Also, `DeleteSource::System` is used here only, which seems pretty redundant, given it's not handled on the receiver side either?


`DeleteSource::System` is intended to notify other components that the history has changed. However, since `Recent Projects` currently calls `DB.recent_projects_on_disk()` on every startup, no component actually relies on this event at the moment.

File: crates/gpui/src/platform.rs [hunk 5]
```diff
@@ -191,6 +191,9 @@ pub(crate) trait Platform: 'static {
     fn set_dock_menu(&self, menu: Vec<MenuItem>, keymap: &Keymap);
     fn perform_dock_menu_action(&self, _action: usize) {}
     fn add_recent_document(&self, _path: &Path) {}
+    fn update_jump_list(&self, _entries: &[&Vec<String>]) -> Option<Vec<Vec<String>>> {
```
[the remark this answers is not in the document]
Comment (JunkuiZhang) [reply crates/gpui/src/platform.rs:194]: Here, I wanted to avoid unnecessary copies, which is why the function signature looks quite complex.

File: crates/gpui/src/platform/windows/destination_list.rs [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    unsafe { list.CommitList() }?;
+    Ok(removed)
+}
+
+const PKEY_TITLE: PROPERTYKEY = PROPERTYKEY {
+    fmtid: GUID::from_u128(0xf29f85e0_4ff9_1068_ab91_08002b27b3d9),
+    pid: 2,
+};
+
+const PKEY_LINK_ARGS: PROPERTYKEY = PROPERTYKEY {
```
[the remark this answers is not in the document]
Comment (JunkuiZhang) [reply crates/gpui/src/platform/windows/destination_list.rs:60]: In Microsoft's official sample, only `PKEY_Title` is used. However, `PKEY_Link_Arguments` is included based on Microsoft's recommendation:  
[Microsoft Docs – IShellLinkW::GetArguments](https://learn.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ishelllinkw-getarguments#remarks)

I copied these two values from `windows-rs`, but I’d prefer not to enable an entire feature of `windows-rs` just for them, as that would introduce a significant amount of additional code.

[PKEY_Title](https://github.com/microsoft/windows-rs/blob/0fc3c2e5a13d4316d242bdeb0a52af611eba8bd4/crates/libs/windows/src/Windows/Win32/Storage/EnhancedStorage/mod.rs#L1881)

[PKEY_Link_Arguments](https://github.com/microsoft/windows-rs/blob/0fc3c2e5a13d4316d242bdeb0a52af611eba8bd4/crates/libs/windows/src/Windows/Win32/Storage/EnhancedStorage/mod.rs#L1525)

File: crates/workspace/src/workspace.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply crates/workspace/src/workspace.rs:4695]: My point is that `HistoryManager` is a singleton behind Arc-like pointer thing already, so scalable as it can get without events.
But not insisting.

File: crates/gpui/src/app.rs [hunk 7]
```diff
@@ -1439,6 +1439,11 @@ impl App {
         self.platform.add_recent_document(path);
     }
 
+    /// Updates the jump list with the updated list of recent paths for the application, only used on Windows for now.
+    pub fn update_jump_list(&self, entries: &[&Vec<String>]) -> Option<Vec<Vec<String>>> {
```
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply crates/gpui/src/app.rs:1443]: I mean that there are two components shared in every OS (Linux is special but has similar concepts in certain DEs, but let's omit it as not implemented in the repo).

* in the OS, there's a list of entries, ordered by "last opened"
  * that list could be altered from the OS side (different ways to do that but still)
* there's a list of last opened projects in Zed, and that list can be altered from Zed and needs to be propagated to the OS side

At let's call all these entries, lists, whatnot similarly, as of now there's "jump_list" for Windows only and something else for mac-only.
What also would be good, is to keep the same, single "add an entry"/"set entries" workflow in the API if possible.

File: crates/gpui/src/platform.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (notmd) [reply crates/gpui/src/platform.rs:194]: Empty vec doesn't allocate. So you can use an empty vec for the None variant.

[the remark this answers is not in the document]
Comment (JunkuiZhang) [reply crates/gpui/src/platform.rs:194]: > Empty vec doesn't allocate. So you can use an empty vec for the None variant.

Good to know! Thanks!

File: crates/gpui/src/platform/windows/destination_list.rs [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    dock_menus: &[DockMenuItem],
+) -> anyhow::Result<Option<Vec<Vec<String>>>> {
+    let (list, removed) = create_destination_list()?;
+    add_recent_folders(&list, entries, removed.as_ref())?;
+    add_dock_menu(&list, dock_menus)?;
+    unsafe { list.CommitList() }?;
+    Ok(removed)
+}
+
+const PKEY_TITLE: PROPERTYKEY = PROPERTYKEY {
```
Comment (mikayla-maki) [inline crates/gpui/src/platform/windows/destination_list.rs:55]: let's include permalinks to where these numbers come from

File: crates/gpui/src/platform/windows/destination_list.rs [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+    entries: &[&Vec<String>],
+    removed: Option<&Vec<Vec<String>>>,
+) -> anyhow::Result<()> {
+    unsafe {
+        let tasks: IObjectCollection =
+            CoCreateInstance(&EnumerableObjectCollection, None, CLSCTX_INPROC_SERVER)?;
+
+        for folder_path in entries
+            .iter()
+            .filter(|path| !is_item_in_array(path, removed))
```
Comment (mikayla-maki) [inline crates/gpui/src/platform/windows/destination_list.rs:121]: It looks like the interior `&Vec<String>` is being treated like a path? Is there any reason to not use a `&Path` here?

File: crates/workspace/src/history_manager.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (mikayla-maki) [reply crates/workspace/src/history_manager.rs:58]: The db wrapper we use, sqlez, does write serialization under the hood. So these writes are controlled several layers down

File: crates/workspace/src/history_manager.rs [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+                WORKSPACE_DB.delete_workspace_by_id(*id).await.log_err();
```
Comment (mikayla-maki) [inline crates/workspace/src/history_manager.rs:59]: I don't think we should remove these from the workspace DB just because the user removed them from the task bar. Is this how VSCode's recent project tracking works on windows?

**Review by mikayla-maki — CHANGES_REQUESTED** [review]

**Comment by mikayla-maki** [comment]
I think the basic functioning of this feature needs to be examined. Conceptually, I don't think this feature should be so deeply integrated with the recent projects and workspace database, and rather it should be the History Manager's job to retain it in memory. So, questions I'd like to answer before we can merge this:

1. Do we need this PR? If the windows behavior of only showing open files is expected behavior across similar apps (e.g. VSCode or Sublime Text), then we don't need all of this extra complexity. 
2. If we do need the PR, is there a 1:1 correspondence between this menu and the recent projects in our database? It feels like there shouldn't be, and this should be a simple "push the last thing you opened into the list". But I don't actually know the semantics here.

Once we've answered those two question, I think we can refactor this to make it a lot simpler.

File: crates/gpui/src/app.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (JunkuiZhang) [reply crates/gpui/src/app.rs:1445]: I don't know if macOS supports setting something like `ssh://...` to the folder path.

Windows menus can not be updated one at a time

File: crates/workspace/src/history_manager.rs (hunk 10 above)
Comment (JunkuiZhang) [reply crates/workspace/src/history_manager.rs:59]: VSCode also removes deleted folders here. According to their comments, if we re-add folders that users have removed back to Recent Folders, Windows gets confused and will refuse to show the entire Recent Folders list.

**Comment by JunkuiZhang** [comment]
> I think the basic functioning of this feature needs to be examined. Conceptually, I don't think this feature should be so deeply integrated with the recent projects and workspace database, and rather it should be the History Manager's job to retain it in memory. So, questions I'd like to answer before we can merge this:
> 
> 1. Do we need this PR? If the windows behavior of only showing open files is expected behavior across similar apps (e.g. VSCode or Sublime Text), then we don't need all of this extra complexity.

Actually, VSCode displays both `Recently Files` AND `Recent Folders`, and keeps them synchronized with its database.

> 2. If we do need the PR, is there a 1:1 correspondence between this menu and the recent projects in our database? It feels like there shouldn't be, and this should be a simple "push the last thing you opened into the list". But I don't actually know the semantics here.
> 
> Once we've answered those two question, I think we can refactor this to make it a lot simpler.

The key thing is that Windows provides a specific mechanism where:

1. Users can delete recent folder items without even opening the app
2. The app only gets notified of deletions when it launches and calls `LoadList`
3. If we include deleted items in our list, Windows will refuse to show the entire `Recent Folders` section

This explains why we need special handling for the Windows implementation.

File: crates/gpui/src/platform/windows/destination_list.rs [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+                action,
+            }),
+            _ => Err(anyhow::anyhow!(
+                "Only `MenuItem::Action` is supported for dock menu on Windows."
+            )),
+        }
+    }
+}
+
+// https://github.com/microsoft/Windows-classic-samples/blob/main/Samples/Win7Samples/winui/shell/appshellintegration/RecipePropertyHandler/RecipePropertyHandler.cpp
```
[the remark this answers is not in the document]
Comment (JunkuiZhang) [reply crates/gpui/src/platform/windows/destination_list.rs:43]: Done

File: crates/workspace/src/history_manager.rs (hunk 10 above)
Comment (mikayla-maki) [reply crates/workspace/src/history_manager.rs:54]: Great, thanks for telling me the platform norms :)

File: crates/gpui/src/app.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (mikayla-maki) [reply crates/gpui/src/app.rs]: Thanks for the explanation :)

**Comment by mikayla-maki** [comment]
Great, thanks for the explanation @JunkuiZhang.

File: crates/workspace/src/workspace.rs [hunk 12]
```diff
@@ -4551,22 +4565,38 @@ impl Workspace {
...
         }
     }
 
-    fn serialize_workspace_internal(&self, window: &mut Window, cx: &mut App) -> Task<()> {
+    /// Set the parameter `update` to `true` if the action changes the history of the workspace.
+    /// Typically, there are three cases that change the history:
+    /// 1. A brand new workspace is opened, or an existing workspace has been deleted.
+    /// 2. An existing workspace is loaded, this will change the order of the workspaces in the history.
+    /// 3. `AddFolderToProject` or `RemoveFromProject` action is triggered.
+    fn serialize_workspace_internal(
```
Comment (mikayla-maki) [inline crates/workspace/src/workspace.rs:4594]: I think this 'update' parameter seems pretty sketchy here. Why exactly do we need to record whether this save updates the history?

**Review by mikayla-maki — CHANGES_REQUESTED** [review]

File: crates/gpui/src/platform/test/platform.rs [hunk 13]
```diff
@@ -360,7 +361,13 @@ impl Platform for TestPlatform {
     }
 
     fn set_menus(&self, _menus: Vec<crate::Menu>, _keymap: &Keymap) {}
-    fn set_dock_menu(&self, _menu: Vec<crate::MenuItem>, _keymap: &Keymap) {}
+    fn set_dock_menu(
+        &self,
+        _menu: Vec<crate::MenuItem>,
+        _keymap: &Keymap,
+    ) -> Vec<SmallVec<[PathBuf; 2]>> {
```
Comment (mikayla-maki) [inline crates/gpui/src/platform/test/platform.rs:368]: I'd rather not change this API if we can. A `Vec<SmallVec<PathBuf>>` return type is a huge increase in complexity for what should be a pretty simple operation. If we have to have a `Vec<Vec<PathBuf>>`, I'd rather that be relegated to a windows-only API, than add it to the general purpose API.

**Review by mikayla-maki — CHANGES_REQUESTED** [review]
PR is looking good! One note on the GPUI API for dock menus, and we should be good to go :)

**Review by mikayla-maki — APPROVED** [review]
Thanks for adding this!
